import "ydb/core/protos/flat_scheme_op.proto";
import "ydb/core/scheme/protos/type_info.proto";

package NTabletFlatScheme;
option java_package = "ru.yandex.kikimr.proto";

message TAlterRecord {
    enum EDeltaType {
        AddTable = 1;
        DropTable = 2;
        AddColumn = 3;
        DropColumn = 4;
        AddColumnToKey = 5;
        AddColumnToFamily = 6;
        AddFamily = 7;
        UpdateExecutorInfo = 8;
        SetCompactionPolicy = 9;
        SetRoom = 10;
        SetFamily = 11;
        SetRedo = 12;
        SetTable = 13;
    };

    required EDeltaType DeltaType = 1;
    optional uint32 TableId = 2;
    optional string TableName = 3;
    optional uint32 ColumnId = 4;
    optional uint32 FamilyId = 7;
    optional uint32 RoomId = 8;

    //_ Global: DB redo log tweaks settings

    optional uint32 Annex = 16;     // Same as Large, but for redo log

    //_ Table: Common options

    optional uint32 ByKeyFilter = 17 [default = 0]; // TPart's key filter
    optional bool EraseCacheEnabled = 20;
    optional uint32 EraseCacheMinRows = 21;
    optional uint32 EraseCacheMaxBytes = 22;
    optional bool ColdBorrow = 23 [default = false];

    //_ Table: Attributes of storage (page collection)

    optional uint32 Main = 9;       // Default channel for main data
    optional uint32 Blobs = 15;     // Channel for external blobs
    optional uint32 Outer = 19;     // Channel for outer packed values

    //_ Table: Attributes of family columns

    optional uint32 Codec = 11;     // Read NPage::ECodec for details
    optional bool InMemory = 12;    // Deprecated, replaced with Cache
    optional uint32 Cache = 13;     // Read NPage::ECache for details
    optional uint32 Large = 14;     // When keep values in single blobs
    optional uint32 Small = 18;     // When pack values to outer blobs

    //_ Table: Attributes of single colimn

    optional string ColumnName = 5;
    optional uint32 ColumnType = 6;
    optional NKikimrProto.TTypeInfo ColumnTypeInfo = 25;
    optional bytes Default = 10;    // Serialized default value for cell
    optional bool NotNull = 24 [default = false];

    optional uint64 ExecutorCacheSize = 100;
    optional NKikimrSchemeOp.TCompactionPolicy CompactionPolicy = 101;
    optional bool ExecutorAllowLogBatching = 102;
    optional uint64 ExecutorLogFlushPeriod = 103;
    optional uint32 ExecutorLimitInFlyTx = 104;
    optional string ExecutorResourceProfile = 105;
    optional bool ExecutorLogFastCommitTactic = 106;
};

message TSchemeChanges {
    repeated TAlterRecord Delta = 1;
};
